File: 1579. Remove Max Number of Edges to Keep Graph Fully Traversable/Solution.java
class Solution {
    public int maxNumEdgesToRemove(int n, int[][] edges) {
        UFDS a = new UFDS(n), b = new UFDS(n);
        int res = 0;
        for (int[] e : edges) {
            if (e[0] != 3) continue;
            if (a.same(e[1], e[2])) res++;
            a.union(e[1], e[2]);
            b.union(e[1], e[2]);
        }
        if (a.sets == 1) return edges.length - n + 1;
        for (int[] e : edges) {
            if (e[0] == 1) {
                if (a.same(e[1], e[2])) res++;
                a.union(e[1], e[2]);
            } else if (e[0] == 2) {
                if (b.same(e[1], e[2])) res++;
                b.union(e[1], e[2]);
            }
        }
        return (a.sets > 1 || b.sets > 1) ? -1 : res;
    }
}

class UFDS {
    int[] parent, rank;
    int sets;

    public UFDS(int n) {
        parent = new int[n + 1];
        rank = new int[n + 1];
        sets = n;
        for (int i = 0; i < n + 1; i++) parent[i] = i;
    }

    public int find(int x) {
        if (parent[x] != x) parent[x] = find(parent[x]);
        return parent[x];
    }

    public boolean same(int x, int y) {
        return find(x) == find(y);
    }

    public void union(int x, int y) {
        if (same(x, y)) return;
        int a = find(x), b = find(y);
        if (rank[a] > rank[b]) {
            parent[b] = a;
        } else {
            parent[a] = b;
            if (rank[a] == rank[b]) rank[b]++;
        }
        sets--;
    }
}

File: 1579. Remove Max Number of Edges to Keep Graph Fully Traversable/Solution.py
class Solution:
    def maxNumEdgesToRemove(self, n: int, edges: List[List[int]]) -> int:
        a, b, res = UFDS(n), UFDS(n), 0
        for i, x, y in edges:
            if i != 3: continue
            if a.same(x, y): res += 1
            a.union(x, y)
            b.union(x, y)
        if a.sets == 1: return len(edges) - n + 1
        for i, x, y in edges:
            if i == 1:
                if a.same(x, y): res += 1
                a.union(x, y)
            elif i == 2:
                if b.same(x, y): res += 1
                b.union(x, y)
        return -1 if a.sets > 1 or b.sets > 1 else res

class UFDS:
    def __init__(self, n):
        self.parent = [i for i in range(n + 1)]
        self.rank = [0 for i in range(n + 1)]
        self.sets = n
    
    def find(self, x):
        if self.parent[x] != x: self.parent[x] = self.find(self.parent[x])
        return self.parent[x]
    
    def same(self, x, y):
        return self.find(x) == self.find(y)
    
    def union(self, x, y):
        if self.same(x, y): return
        a, b = self.find(x), self.find(y)
        if self.rank[a] > self.rank[b]:
            self.parent[b] = a
        else:
            self.parent[a] = b
            if self.rank[a] == self.rank[b]: self.rank[b] += 1
        self.sets -= 1